Advanced flow (#52)
openid/OpenID4VP: openid-4-verifiable-presentations-1_0.md

Implementations can use any pre-existing OAuth 2.0 Grant Type and Response Type in conjunction with this specification to support different deployment architectures.

> OpenID for Verifiable Presentations supports scenarios where the Authorization Request is sent both when the Verifier is interacting with the End-User using the device that is the same or different from the device on which requested Credential(s) are stored. It also allows the Verifier to first discover the Wallet's capabilities before the actual presentation request is created and passed to the wallet. 

Reviewer: ```suggestion
OpenID for Verifiable Presentations supports scenarios where the Authorization Request is sent both when the Verifier is interacting with the End-User using the device that is the same or different from the device on which requested Credential(s) are stored. 

This specification also provides the Verifier an option to first discover the Wallet's capabilities before the actual presentation request is created and passed to the Wallet. 
```

this point is significant enough it should have a separate paragraph.

Implementations can use any pre-existing OAuth 2.0 Grant Type and Response Type in conjunction with this specification to support different deployment architectures.

> OpenID for Verifiable Presentations supports scenarios where the Authorization Request is sent both when the Verifier is interacting with the End-User using the device that is the same or different from the device on which requested Credential(s) are stored. 

Reviewer 2: yes, let's try to make it simpler.

> OpenID for Verifiable Presentations supports scenarios where the Authorization Request is sent both when the Verifier is interacting with the End-User using the device that is the same or different from the device on which requested Credential(s) are stored. 

Author: suggestions welcome

> OpenID for Verifiable Presentations supports scenarios where the Authorization Request is sent both when the Verifier is interacting with the End-User using the device that is the same or different from the device on which requested Credential(s) are stored. 

Reviewer 3: ```suggestion
OpenID for Verifiable Presentations supports the scenario in which the Verifier sends an Authorization Request to  an End-User using either the same or different device from which the requested Credential(s) are stored. 
```

how's this?

(3) The Wallet prepares the Verifiable Presentation(s) of the Verifiable Credential(s) that the End-User has consented to. It then sends to the Verifier an Authorization Response where the Verifiable Presentation(s) are contained in the `vp_token` parameter.

> ## Advanced Flow with Wallet Capabilities Discovery {#adv_flow}

Reviewer: ```suggestion
## Cross Device Flow with Wallet Capabilities Discovery {#adv_flow}
```
Not sure I am a big fan of "advance flow" term - feels too abstract.

Reviewer 2: Yes, I'd tend to not use comparative terms like advanced or High or anything that might produce a relation or reference to something lower or feature-less

Reviewer: I think the way this section is written right now does not exactly convey why this is an "advanced" flow. maybe i am misunderstanding something but I would expect more focus on the new ability for the wallet to communicate its capabilities/endpoint/attestation

Author: Advanced is the title of the PR, now it seems it will be a discovery (pre-step). Changed title and first sentence.

> Below is a diagram of a flow where the End-User presents a Credential to a Verifier interacting with the End-User on a different device as the device the Wallet resides on. In this case, the Verifier starts the flow by discovering the Wallet's capabilities through an additional message exchange. This is especially useful, if the Verifier starts the process using a URL serving as alias for a group of wallets (e.g. a custom scheme) but would like to tailor the presentation request to the particular wallet. 

Reviewer: ```suggestion
Below is a diagram of a flow where the Verifier starts the flow by discovering the Wallet's capabilities through an additional message exchange. This is especially useful, if the Verifier starts the process using a URL serving as alias for a group of Wallets (e.g. a custom scheme) but would like to tailor the presentation request to the particular Wallet. This example uses cross-device flow where the End-User presents a Credential to a Verifier interacting with the End-User on a different device as the device the Wallet resides on. 
```
I would emphasize the wallet capability selection part as opposed to starting with cross-device description.

Reviewer 4: "different device as the device the Wallet resides on" -> "different device than the device that the Wallet resides on" maybe reads more clearly (or maybe it's just me).

Author: I made a mistake, as the flow obviously is same device. 
The sentence (and with that the order of things) was copied from the other flows. It just spells out the fact that this is a same device flow. 
Changed the text to refer to the same device flow instead of duplicating text.

Author: please check

> In this flow, the Verifier first prepares a discovery request and sends it to the Wallet. The Wallet answers this request by sending an HTTP POST request to the Verifier, providing the Verifier with its Issuer URL and further parameters. The Verifier then creates a presentation request tailored to the capabilities and trust domain of the partiular Wallet and passes the request object in the HTTP POST response. The rest of the flow works like described in (#same_device). The discovery request can also be used in cross device scenarios, the Verifier would render the disovery message in a QR code.  

Reviewer: +1 I had the same feedback as Giuseppe. probably something like `providing the Verifier with an app link/universal link specific to that Wallet`. 
though is it only app/universal link that the Wallet can return? does it make sense returning cloud wallet's backend endpoint?

> In this flow, the Verifier first prepares a discovery request and sends it to the Wallet. The Wallet answers this request by sending an HTTP POST request to the Verifier, providing the Verifier with its Issuer URL and further parameters. The Verifier then creates a presentation request tailored to the capabilities and trust domain of the partiular Wallet and passes the request object in the HTTP POST response. The rest of the flow works like described in (#same_device). The discovery request can also be used in cross device scenarios, the Verifier would render the disovery message in a QR code.  

Reviewer 4: also unsure what "Issuer URL" means here

> In this flow, the Verifier first prepares a discovery request and sends it to the Wallet. The Wallet answers this request by sending an HTTP POST request to the Verifier, providing the Verifier with its Issuer URL and further parameters. The Verifier then creates a presentation request tailored to the capabilities and trust domain of the partiular Wallet and passes the request object in the HTTP POST response. The rest of the flow works like described in (#same_device). The discovery request can also be used in cross device scenarios, the Verifier would render the disovery message in a QR code.  

Author: The Issuer URL ist the URL identifying the Wallet as an AS. This URL is used to fetch the wallet's metadata.

Note: The diagram does not illustrate all the optional features of this specification.

Note: The usage of the Request URI as defined in [@!RFC9101] does not depend on any other choices made in the protocol extensibility points, i.e., it can be used in the Same Device Flow, too.

!---
~~~ ascii-art
~~~ ascii-art
+--------------+   +--------------+                                    +--------------+
|     User     |   |   Verifier   |                                    |    Wallet    |
+--------------+   +--------------+                                    +--------------+  
        |                 |                                                   |
        |    Interacts    |                                                   |
        |---------------->|                                                   |
        |                 |  (1) Discovery Request                            |
        |                 |  (Presentation Request URI)                       |
        |                 |-------------------------------------------------->|
        |                 |                                                   |

>         |                 |  (2) Request the Request Object                   |

Reviewer: I think it should be clearer that this (2) is also a "Discovery Response", correct?

Reviewer: ```suggestion
        |                 |  (2) Discovery Response                                    |
```

Author: there is discovery, but the result is a request object

Reviewer 2: this reminds me the italian flow where we have mixed the provisioning of the wia with DPoP to the request_uri endpoint.

We are looking for a better specialization of the endpoints, then the proposed, alternative, flow was provided in this comment: https://github.com/openid/OpenID4VP/pull/52#issuecomment-1741179550

        |                 |<--------------------------------------------------|
        |                 |                                                   |
        |                 |  (2.5) Respond with the Request Object            |
        |                 |      (Presentation Definition)                    |
        |                 |-------------------------------------------------->|
        |                 |                                                   |
        |   User Authentication / Consent                                     |
        |                 |                                                   |
        |                 |  (3)   Authorization Response                     |
        |                 |  (VP Token with Verifiable Presentation(s))       |
        |                 |<--------------------------------------------------|
~~~
!---

> Figure: Cross Device Flow

Reviewer: ```suggestion
Figure: Wallet Discovery Flow
```

Author: changed to "Same Device Flow with Wallet Capabilities Discovery"

Reviewer 2: I would like to cover both cross device and same device within this PR

an alternative flow was provided aiming to satisfy this: https://github.com/openid/OpenID4VP/pull/52#issuecomment-1741179550

> (1) The Verifier sends to the Wallet a Discovery Request that contains a Presentation URI from where to obtain the Request Object containing the actual Authorization Request object. 

Reviewer: ```suggestion
(1) The Verifier sends to the Wallet a Discovery Request that contains a Presentation URI to where Wallet can send information about itself and obtain the Request Object containing the actual Authorization Request. 
```

> (2) The Wallet sends an HTTPS POST request to the Presentation URI to retrieve the Request Object.

Reviewer: ```suggestion
(2) The Wallet sends an HTTPS POST request to the Presentation URI that can contain information about wallet's endpoint, capabilities and attestation.
```

Reviewer 2: wallet's endpoint -> Wallet's endpoint

Author: I adopted the text to convey what I perceive was the objective of the proposal, I used different text since there is no capabilities nor endpoint directly be conveyed in the POST request as it is specified now.

> (2.5) The HTTPS POST response returns the Request Object containing Authorization Request parameters. It especially contains a Presentation Definition as defined in [@!DIF.PresentationExchange] that describes the requirements of the Credential(s) that the Verifier is requesting to be presented. Such requirements could include what type of Credential(s), in what format(s), which individual Claims within those Credential(s) (Selective Disclosure), etc. The Wallet processes the Request Object and determines what Credentials are available matching the Verifier's request. The Wallet also authenticates the End-User and gathers her consent to present the requested Credentials. 

Reviewer: ```suggestion
(2.5) The HTTPS POST response returns the Request Object containing Authorization Request parameters based on the information specific to the Wallet that it received in step (2). It especially contains a Presentation Definition as defined in [@!DIF.PresentationExchange] that describes the requirements of the Credential(s) that the Verifier is requesting to be presented. Such requirements could include what type of Credential(s), in what format(s), which individual Claims within those Credential(s) (Selective Disclosure), etc. The Wallet processes the Request Object and determines what Credentials are available matching the Verifier's request. The Wallet also authenticates the End-User and gathers her consent to present the requested Credentials. 
```
do we need to repeat `it especially contains...` part which is duplicative across all examples? guess we do in case reader start reading one example without reading others..

(3) The Wallet prepares the Verifiable Presentation(s) of the Verifiable Credential(s) that the End-User has consented to. It then sends to the Verifier an Authorization Response where the Verifiable Presentation(s) are contained in the `vp_token` parameter.

> ## Same Device Flow with Wallet Capabilities Discovery {#adv_flow}

Reviewer: I think {#xxx_xxx} shows in the URL to this section, so I think we should start cleaning it up something like {#capability_discovery_flow}. (this can be a separate issue/PR)

Author: I changed the label of this section

> Below is a diagram shows a flow a describe in ({#same_device}) with an additional pre-step used to discover the Wallet's capabilities and tailor the actual request to it.

Reviewer: ```suggestion
Below is a diagram that shows a flow described in (#same_device), but with an additional pre-step used that allows the Verifier to discover the Wallet's capabilities and tailor the actual request.
```

> In this case, the Verifier starts the flow by discovering the Wallet's capabilities through an additional message exchange. This is especially useful, if the Verifier starts the process using a URL serving as alias for a group of wallets (e.g. a custom scheme) but would like to tailor the presentation request to the particular wallet. 

Reviewer 5: We should add the following Note somewhere
Note. In order to prevent a Verifier from fingerprinting the wallet, it is strongly recommended that one or more standardised sets of capabilities are defined that may be commonly supported by wallets. This will allow the Verifier to request conformance to one or more sets, and the wallet to return an assertion about one or more sets, thereby protecting the privacy of the wallet implementation.

> In this case, the Verifier starts the flow by discovering the Wallet's capabilities through an additional message exchange. This is especially useful, if the Verifier starts the process using a URL serving as alias for a group of wallets (e.g. a custom scheme) but would like to tailor the presentation request to the particular wallet. 

Reviewer 2: I agree, since there may be cases where the wallet instances doesn't support discovery, then the RP has to rely on a standard set. 

@David-Chadwick do you think htat this set should be recommended in the current spec or that it would be out of scope and then it should be defined within each impl profile? (eg: EUDI Wallet should have its standard set ... and so on)

> In this case, the Verifier starts the flow by discovering the Wallet's capabilities through an additional message exchange. This is especially useful, if the Verifier starts the process using a URL serving as alias for a group of wallets (e.g. a custom scheme) but would like to tailor the presentation request to the particular wallet. 

Reviewer 6: Since David is not concerned about interoperability here, but privacy, this could go into the privacy considerations I mentioned in my comment on this PR. However, I think that just using some set of default capabilities is not sufficient: There are many, many ways in which information about the wallet can be leaked. Browser vendors are working quite hard to make browser fingerprinting harder, and we need to take a page or two out of their book if we allow this request to happen without user consent, to implement similar protections. For example: Just the presence of a particular wallet might leak the nationality of a user, and that particular wallet might send parameters in a different order than all other wallets. Or: The wallet might support a different version of HTTP than all other wallets.

> In this case, the Verifier starts the flow by discovering the Wallet's capabilities through an additional message exchange. This is especially useful, if the Verifier starts the process using a URL serving as alias for a group of wallets (e.g. a custom scheme) but would like to tailor the presentation request to the particular wallet. 

Reviewer 5: I agree that privacy protection is extremely difficult (and maybe its impossible for it to be perfect) but we should at least add guidelines or recommendations into how the capabilities detection feature should be used/implemented.
I am happy for my proposed text to go into the privacy considerations section, supplemented with additional cautionary text from Daniel.
@peppelinux Certainly profiles of the spec will define their own set of required capabilities for all implementors of the profile, but if the WG determine that there is a useful common subset that is widely applicable, this could go into an informative annex. (An analogy is the X.500 spec that defined a common set of useful attributes that most people implemented, whilst acknowledging that other groups of users would define their own additional attributes.)

> In this case, the Verifier starts the flow by discovering the Wallet's capabilities through an additional message exchange. This is especially useful, if the Verifier starts the process using a URL serving as alias for a group of wallets (e.g. a custom scheme) but would like to tailor the presentation request to the particular wallet. 

Reviewer 2: @danielfett by tracking Wallet Instance IPv4 with GeoIP it would be enough to track nationality and also mobile broadband provider

> In this case, the Verifier starts the flow by discovering the Wallet's capabilities through an additional message exchange. This is especially useful, if the Verifier starts the process using a URL serving as alias for a group of wallets (e.g. a custom scheme) but would like to tailor the presentation request to the particular wallet. 

Reviewer 2: @danielfett even if a default capabilities is not sufficient (I agree) specs like OIDC requires the sigalg RS256 as a strong recommendation to the entities, SAML2 made something similar in the real-world with sha-1 and then SHA256. I mentions these just to continuing debating on the requirement of giving more guidance to implementers

> In this case, the Verifier starts the flow by discovering the Wallet's capabilities through an additional message exchange. This is especially useful, if the Verifier starts the process using a URL serving as alias for a group of wallets (e.g. a custom scheme) but would like to tailor the presentation request to the particular wallet. 

Author: Please suggest text that I can add to the privacy considerations section.

> In this flow, the Verifier first prepares a discovery request to be provided to the Wallet. The Wallet answers this request by sending an HTTP POST request to the Verifier, providing the Verifier with its Issuer URL (identifying the wallet provider and used to fetch wallet metadata) and further parameters. The Verifier then creates a presentation request tailored to the capabilities and trust domain of the particular Wallet and passes the request object in the HTTP POST response. The rest of the flow works like described in (#same_device). The discovery request can also be used in cross device scenarios where the Verifier would render the discovery message in a QR code.  

Author: the verifier receives a request via HTTPS, I don't see how this works for offline scenarios

        |                 |-------------------------------------------------->|
        |                 |                                                   |
        |                 |  (2) Request the Request Object                   |
        |                 |<--------------------------------------------------|
        |                 |                                                   |
        |                 |  (2.5) Respond with the Request Object            |
        |                 |      (Presentation Definition)                    |
        |                 |-------------------------------------------------->|
        |                 |                                                   |
        |   User Authentication / Consent                                     |
        |                 |                                                   |
        |                 |  (3)   Authorization Response                     |
        |                 |  (VP Token with Verifiable Presentation(s))       |
        |                 |<--------------------------------------------------|
~~~
!---
Figure: Cross Device Flow

> (1) The Verifier sends to the Wallet a Discovery Request that contains a Presentation URI to where Wallet can send information about itself and obtain the Request Object containing the actual Authorization Request. 

Reviewer 2: @David-Chadwick ^

> (1) The Verifier sends to the Wallet a Discovery Request that contains a Presentation URI to where Wallet can send information about itself and obtain the Request Object containing the actual Authorization Request. 

Author: This level of detail is used in same device section, too. I suggest to file an issue to change both.

Presentation of Verifiable Credentials using OpenID for Verifiable Presentations can be combined with the user authentication using [@SIOPv2], and the issuance of OAuth 2.0 Access Tokens.

# Discovery Request

> The Discovery Request allows Verifers to discover the Wallet's capabilities before they create the actual presentation request. It also allows the Wallet to authenticate towards the Verifier and to provide the Verifier with additional data used in the creation of the request signature and to encrypt the request at the application layer (if needed).

Reviewer: ```suggestion
The Discovery Request allows Verifers to discover the Wallet's capabilities before they create the actual presentation request. It also allows the Wallet to authenticate towards the Verifier and to provide the Verifier with additional data used in the creation of the request signature and to encrypt the request at the application layer (if needed).
```
in my mind, wallet's endpoint, wallet's capabilities, and wallet attestation are three different things, and I am not sure "wallet capabilities" can be used to express them all

Reviewer 2: I assume that the wallet instance attestation contains the wallet's capabilities, materially defined with:

- cnf.jwk (any other cnf.*)
- NiST's AAL (the equivalence of the italian `attested_security_contexts`)
- Authorization Server metadata
- other wallet instance metadata, as client_metadata when a wallet instance acts like a Verifier.

Author: those three things are differentiated in the current text. wallet capabilities is in the first sentence, authentication in the second sentence

Author: @peppelinux I assume the capabilities are conveyed using wallet metadata. I agree, there can be other ways. Let's discuss how we can support them.

> The Discovery Request message has the following parameters:

Reviewer: ```suggestion
This specification defines the following new parameters for the Discovery Request message:
```

> `presentation_request_uri`:
> : A string containing an HTTPS URL pointing to a resource under the control of the Verifier where the Wallet is supposed to obtain the presentation request object. 

Reviewer: I don't think `presentation_request_uri` parameter name could be more explicit and differentiate it more clearly from `request_uri` in rfc9101..? something like `discovery_and_request_uri` (yes, I know i am not brilliant at names)?

I would also borrow some language from https://www.rfc-editor.org/rfc/rfc9101#section-5-2.4 like 
```suggestion
`presentation_request_uri`:
: The absolute URI, as defined by [RFC 3986](https://www.rfc-editor.org/rfc/rfc9101#RFC3986) [[RFC3986](https://www.rfc-editor.org/rfc/rfc9101#RFC3986)], that receives data from the Wallet such as endpoint, capabilities and attestation and returns the [Request Object](https://www.rfc-editor.org/rfc/rfc9101#name-request-object) containing the authorization request parameters defined in [Section 5](https://openid.bitbucket.io/connect/openid-4-verifiable-presentations-1_0.html#section-5).
```
not sure we need ` If this parameter is present in the authorization request, request MUST NOT be present.`

Reviewer 2: I'm not in favor of discovery AND request, since I assume that there might be cases where a wallet solution doesn't support the discovery feature then the request_uri endpoint should be untouched

I assume, according to my implementations, that the RP that can't assert the wallet capabilities should attests to the wallet instance the lowest level of security, if supported for the requirements of the LoA

Author: The idea in the PR is to have a new endpoint (the presentation endpoint) that bundles discovery and request creation. We could also split this into discovery and request_uri. However I would like to understand the consequences. What directly comes to mind is that the state must be kept at the verifier between the two requests whereas the current design allows the verifier to create the request and directly release it to the wallet.

Reviewer 2: I'm for `discovery_response_uri` where the HTTP response could be 302 Found with the discovery_state that binds the discovery to the request issuance

then we have divided the endpoints by using the request_uri endpoint with the sole addition of the discovery_state that allows the RP to do its lookups. This value could be also encrypted (depending by what would cost more in an infrastructure: CPU or Storage!)

`context`:

> : A string identifying the context of the discovery request from the perspective of the Verifier. The value is opqaue to the Wallet, it MUST pass this value to the Presentation Endpoint Request (see below). 

Author: I intentionally did not designate it as state as it is intended to be used to label the context ("entrance to east wing") and not carry a certain state of a certain transaction. This allows to use the same discovery request for multiple transactions (print it out and put it onto the wall).

> : A string identifying the context of the discovery request from the perspective of the Verifier. The value is opqaue to the Wallet, it MUST pass this value to the Presentation Endpoint Request (see below). 

Reviewer 2: I fully agree with Kristina, `context` is something too much related to other specs and schemas.
I propose `discovery_state` with the will to keep the words always on the same direction, the good semantic help us to reduce the brain stress

> The Discovery Request is represented as a JSON object, where all parameters are top level JSON claims. 

Reviewer: ```suggestion
The Discovery Request MUST be represented as a JSON object with all parameters as top-level JSON claims. 
```

Reviewer 4: JSON doesn't have claims.

Reviewer 2: according to rfc8259

```
An object structure is represented as a pair of curly brackets
   surrounding zero or more name/value pairs (or members).
```

Author: @bc-pi what do you suggest?

The following is an example request: 

```JSON
{
    "presentation_request_uri": "verifier.example.com/presentation_request",
    "context": "register_kyc"
}

> ```

Reviewer: just to note that parameter names might change based on the comments above.

> The Discovery Request is either send the Discovery Endpoint of the Wallet or it is rendered as a QR Code and scanned with the Wallet or a camera app. 

Reviewer: ```suggestion
The Discovery Request is either sent to the Discovery Endpoint of the Wallet or is rendered as a QR Code.
```
i think reference to the section where discovery endpoint (not sure about the name...) is defined would be nice

> Note: the Discovery Request intentionally does not use OAuth parameters, such as a `client_id`. The idea is to allow the Verifier to select the Client Identifier it wants to use for interacting with the Wallet after it has determined the trust mechanisms and domains the wallet supports.   

Reviewer 2: ```suggestion
Note: the Discovery Request intentionally does not use OAuth parameters, such as a `client_id`. The idea is to allow the Verifier to select the Client Identifier it wants to use for interacting with the Wallet after it has determined the trust mechanisms and domains the Wallet supports.   
```

Reviewer 2: Not in favour of this because the RP must provide it's unique identifier since a wallet solution, or a trust model, may require to not send the wallet instance attestation before having established the trust to the RP

A client_id allows this in both oidc federation and eidas trusted list

Reviewer: i can see both sides, but if the value of this mechanism is for the verifier to be able to tailor the request based on the POST from the wallet, this sentence makes sense. to avoid sending data to random verifiers, wallet can have the list of trusted `presentation_request_uri`? or append .well-known to the presentation_request_uri? (thinking out loud here)

> ## Discovery Endpoint

Reviewer: I think this section needs to be in the Verifier Metadata section https://openid.bitbucket.io/connect/openid-4-verifiable-presentations-1_0.html#name-wallet-metadata-authorizati and not here

Author: This endpoint is not the typical static "get some metadata" endpoint. It is part of the application logic of the wallet. I therefore added it here. I think this is the best solution for readability and understandability.

> This is an HTTPS endpoint offered by the Wallet. The Discovery Request MUST be sent as a HTTPS GET request or an HTTP redirect to the Discovery Endpoint URL defined in (#wallet-metadata)

Reviewer 2: ```suggestion
This is an HTTPS endpoint offered by the Wallet. The Discovery Request MUST be sent as an HTTPS GET request or an HTTP redirect to the Discovery Endpoint URL defined in (#wallet-metadata)
```

Reviewer 4: does {#wallet-metadata} exist anywhere? Seems to be causing problems in the "build" https://github.com/openid/OpenID4VP/actions/runs/6324297591/job/17173526145?pr=52 

```2023/09/27 09:52:29 Empty 'stream' in [seriesInfo], resulting XML may fail to parse.
./openid-4-verifiable-presentations-1_0-20.xml(0): Error: IDREF attribute target references an unknown ID "wallet-metadata", at None
/data/openid-4-verifiable-presentations-1_0-20.xml(3): Error: Invalid document before running preptool.
Unable to complete processing ./openid-4-verifiable-presentations-1_0-20.xml
Error: Process completed with exit code 1.```

> The HTTPS URL contains a single URI query parameter `discovery` where the value is the JSON object as defined above. 

Reviewer 2: ```suggestion
The HTTPS URL MUST contain a single URI query parameter `discovery` where the value is the JSON object as defined above. 
```

> The HTTPS URL contains a single URI query parameter `discovery` where the value is the JSON object as defined above. 

Reviewer: I don't understand this.

## Presentation Endpoint

The Presentation Endpoint is an HTTPS endpoint exposed by the Verifier. 

> ### Presentation Request

Reviewer: I am not comfortable calling this `presentation request`. requests in OID4VP are from the verifier. it is confusing to introduce a `request` coming from the wallet. why can't this be a `discovery response` to `discovey request` from the verifier? or `wallet capability offer` to learn from VCI

Presentation Requests MUST be HTTPS POST requests with the "application/json" media type.

> The follwowing parameters are defined: 

Reviewer: ```suggestion
This specification defines the following new parameters:
```

> The follwowing parameters are defined: 

Reviewer: please indicate which parameters is REQUIRED vs OPTIONAL - hard to review without that

> `context`:
> : A JSON String containing the value of the corresponding Discovery Request's `context` parameter.

Reviewer: ```suggestion
`verifier_state`:
: A JSON String containing the value of the corresponding Discovery Request's `verifier_state` parameter.
```

`issuer`:

> : A JSON containg a HTTPS URL designating the Issuer URL of the Wallet. 

Reviewer 2: ```suggestion
: A JSON containing an HTTPS URL designating the Issuer URL of the Wallet. 
```

Reviewer 2: I didn't get the Issuer URL of the Wallet

Reviewer: +1 same as a thread here https://github.com/openid/OpenID4VP/pull/52/files?diff=unified&w=0#r1339310421

Author: by relying on RFC8414 for wallet metadata (Section 8.1) the wallet as AS needs to have an Issuer URL

> : A JSON containg a HTTPS URL designating the Issuer URL of the Wallet. 

Reviewer: also this PR needs to define how wallet capability/endpoint parameters can be obtained by appending `/.well-known/oauth-authorization-server` to this issuer URL. also I think it should define how wallet can pass those parameters by value in this POST message - i can totally see wallets doing that.

Author: added text re 'oauth-authorization-server' we-known location

we need to discuss the security implications of passing metadata as parameters. I guess that is @peppelinux has some ideas about it

obtaining the data through /.well-known/oauth-authorization-server at least binds the data to the issuer URL

`w_nonce`:
: A JSON String containg a cryptographic nonce the Verifier MUST use when creating the signed presentation request object. 

`client_assertion`

> : A JSON String containing a wallet attestation along with a proof of posession of the confiration key as defined in [@!I-D.ietf-oauth-attestation-based-client-auth]. This assertion is used to authenticate the Wallet towards the Verifier. 

Reviewer 2: ```suggestion
: A JSON String containing a Wallet attestation along with a proof of possession of the configuration key as defined in [@!I-D.ietf-oauth-attestation-based-client-auth]. This assertion is used to authenticate the Wallet towards the Verifier. 
```

Reviewer 4: ietf-oauth-attestation-based-client-auth is a method of client authentication to the AS. I don't think it works to just reference here as a way to authenticate the Wallet (the AS in this context) towards the Verifier (the client in this context).

Author: happy to extend the text if we have basically agreed to use the mechanism

### Presentation Response

Presentation Response MUST be HTTPS POST response with the "application/json" media type.

The follwowing parameters are defined: 

`request`:

> : A JSON String containing the signed request object as defined in [@RFC9101]. It MUST fulfill the requirements as defined in (#vp_token_request).

Reviewer: I think this is confusing. why is there suddenly a need to define a top level claim? usually the entire object in the response is a request object without any top level claims. the parameter name also conflicts with the `request` parameter defined in rfc 9101 to pass the request by value

### Presentation Error Response

> The error code `401` signals to the wallet that it needs to authenticate to the Verifier. In this case, the error response SHOULD contain a `WWW-Authenticate` header for every attestation method the Verifier supports.

Reviewer 2: ```suggestion
The error code `401` signals to the Wallet that it needs to authenticate to the Verifier. In this case, the error response SHOULD contain a `WWW-Authenticate` header for every attestation method the Verifier supports.
```

> This is an example for the wallet attestation method as specified above. The `WWW-Authenticate` contains the nonce value the Wallet MUST use in the calculation of the proof of possession of the respective wallet attestation. 

Reviewer 2: ```suggestion
Below a non-normative example for the Wallet attestation method as specified above. The `WWW-Authenticate` contains the nonce value the Wallet MUST use in the calculation of the proof of possession of the wallet attestation. 
```

# Discovery Request

The Discovery Request allows Verifers to discover the Wallet's capabilities before they create the actual presentation request. It also allows the Wallet to authenticate towards the Verifier and to provide the Verifier with additional data used in the creation of the request signature and to encrypt the request at the application layer (if needed).

This specification defines the following new parameters for the Discovery Request message:

`presentation_request_uri`:
: A string containing an HTTPS URL pointing to a resource under the control of the Verifier where the Wallet is supposed to obtain the presentation request object. 

`interaction`:
: A string identifying the context of the discovery request from the perspective of the Verifier. The value is opqaue to the Wallet, it MUST pass this value to the Presentation Endpoint Request (see below). 

The Discovery Request MUST be represented as a JSON object with all parameters as top-level JSON claims. 

The following is an example request: 

```JSON
{
    "presentation_request_uri": "verifier.example.com/presentation_request",

>     "context": "register_kyc"

Reviewer 4: the parameter name changed from context to interaction, i think? 
```suggestion
    "interaction": "register_kyc"
```

Presentation of Verifiable Credentials using OpenID for Verifiable Presentations can be combined with the user authentication using [@SIOPv2], and the issuance of OAuth 2.0 Access Tokens.

# Discovery Request

The Discovery Request allows Verifers to discover the Wallet's capabilities before they create the actual presentation request. It also allows the Wallet to authenticate towards the Verifier and to provide the Verifier with additional data used in the creation of the request signature and to encrypt the request at the application layer (if needed).

This specification defines the following new parameters for the Discovery Request message:

`presentation_request_uri`:
: A string containing an HTTPS URL pointing to a resource under the control of the Verifier where the Wallet is supposed to obtain the presentation request object. 

`context`:
: A string identifying the context of the discovery request from the perspective of the Verifier. The value is opqaue to the Wallet, it MUST pass this value to the Presentation Endpoint Request (see below). 

> The Discovery Request MUST be represented as a JSON object with all parameters as top-level JSON claims. 

Reviewer 2: ```suggestion
The Discovery Request MUST be a JSON object with all members in the top-level. 
```

The following is an example request: 

```JSON
{

>     "presentation_request_uri": "verifier.example.com/presentation_request",

Reviewer 2: according to my rpevious comments, it would be

`"discovery_response_uri": "https://verifier.example.com/discovery-response"`

    "context": "register_kyc"
}
```

The Discovery Request is either sent to the Discovery Endpoint of the Wallet or is rendered as a QR Code.

Note: the Discovery Request intentionally does not use OAuth parameters, such as a `client_id`. The idea is to allow the Verifier to select the Client Identifier it wants to use for interacting with the Wallet after it has determined the trust mechanisms and domains the Wallet supports.   

## Discovery Endpoint

> This is an HTTPS endpoint offered by the Wallet. The Discovery Request MUST be sent as an HTTPS GET request or an HTTP redirect to the Discovery Endpoint URL defined in (#wallet-metadata).

Reviewer 2: `This is an HTTPS endpoint offered by the Wallet.`

please don't, we should assume that the wallet should not require any online HTTPs endpoint

Author: you are absolutely right, I rewrote the section. please have a look

The HTTPS URL MUST contain a single URI query parameter `discovery` where the value is the JSON object as defined above. 

## Presentation Endpoint

The Presentation Endpoint is an HTTPS endpoint exposed by the Verifier. 

### Presentation Request

Presentation Requests MUST be HTTPS POST requests with the "application/json" media type.

> The follwowing parameters are defined: 

Reviewer 2: ```suggestion
The following parameters are defined: 
```

`context`:
: A JSON String containing the value of the corresponding Discovery Request's `context` parameter.

`issuer`:

> : A JSON containing an HTTPS URL designating the Issuer URL of the Wallet. The Verifier can obtain the Wallet's metadata by adding the well-know location `oauth-authorization-server` as specifid in [@!RFC8414].

Reviewer 2: this term `Issuer URL of the Wallet` or `Issuer URL` is misleading, would it mean Wallet Provider?

I'm strongly not in favour of requiring the entities to be online and that the wallet might exposes its metadata thought the wallet providers endpoints

the wallet instance would be the best place, attestable and not-repudiable, where the wallet metadata can be provided.

`w_nonce`:

> : A JSON String containg a cryptographic nonce the Verifier MUST use when creating the signed presentation request object. 

Reviewer 2: where the reader is able to distinguish between nonce and cryptographic nonce?
I would suggest a ref where the definition of cryptographic nonce is provided and the way to issue it

`client_assertion`

> : A JSON String containing a Wallet attestation along with a proof of possession of the configuration key as defined in [@!I-D.ietf-oauth-attestation-based-client-auth]. This assertion is used to authenticate the Wallet towards the Verifier. 

Reviewer 2: ```suggestion
: A JSON String containing a Wallet attestation along with a proof of possession of the public key as defined in [@!I-D.ietf-oauth-attestation-based-client-auth]. This assertion is used to authenticate the Wallet towards the Verifier. 
```
